apps/src/services/download.ts

```diff
@@ -0,0 +1,157 @@
+): void => {
+	const hasBbox = 'bbox' in payload && payload.bbox !== undefined;
+	const hasPoints = 'points' in payload && payload.points !== undefined;
+
+	if (!hasBbox && !hasPoints) {
+		throw new ExactlyOneGeographicSelectorError(
+			'Download requires either a selected region (bbox) or selected grid cells (points)'
+		);
+	}
+
+	if (hasBbox && hasPoints) {
+		throw new ExactlyOneGeographicSelectorError(
+			'Download cannot have both bbox and points - only one geographic selector allowed'
+		);
+	}
+};
+
+/**
+ * Fetches binary file data from a download endpoint and returns a temporary blob URL.
+ *
+ * Validates the response contains actual file data (not empty or error page) before
+ * creating a browser-usable object URL for triggering downloads.
+ *
+ * @param url - Download endpoint URL (must contain '/download' in pathname)
+ * @param payload - Request data specifying file format and geographic region
+ * @returns Temporary blob URL string for use with download links
+ * @throws {FetchPostDownloadError} On network errors, invalid responses, or empty files
+ */
+export const postDownloadToBlobObjectURL = async (
+	url: string,
+	payload: PostDownloadToBlobObjectURLPayload
+): Promise<string | null> => {
+	const urlObj = new URL(url);
+
+	// Validate URL points to download endpoint
+	if (urlObj.pathname.startsWith('/download') === false) {
+		const message = `The provided URL does not point to a download endpoint: ${url}`;
+		console.error(message);
+		throw new FetchPostDownloadError(message);
+	}
+
+	try {
+		// Send the POST request with the payload as JSON
+		const response = await fetch(url, {
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/json',
+			},
+			body: JSON.stringify(payload),
+		});
+
+		// Throw an error if the response is not OK (4xx or 5xx)
+		if (!response.ok) {
+			throw new FetchPostDownloadError(
+				`HTTP error! status: ${response.status}`
+			);
+		}
+
+		/**
+		 * Example response from download endpoint:
+		 *
+		 * ```http
+		 * POST /download HTTP/1.1
+		 * Content-Type: application/json
+		 *
+		 * HTTP/1.1 200 OK
+		 * Content-Type: application/zip
+		 * Content-Length: 25077
+		 * Content-Disposition: attachment; filename=tx_max.zip
+		 * Cache-Control: max-age=604800
+		 * ```
+		 *
+		 * Note: Content-Disposition is ignored by fetch() - only affects direct browser navigation
+		 */
+
+		// Validate response contains binary file data, not an error page
+		const contentType = response.headers.get('Content-Type');
+		if (
+			!contentType?.includes('application/zip') &&
+			!contentType?.includes('application/octet-stream')
+		) {
+			throw new FetchPostDownloadError(
+				`Unexpected content type: ${contentType}`
+			);
+		}
+
+		// Check response isn't empty before reading stream
+		const contentLength = response.headers.get('Content-Length');
+		if (contentLength && parseInt(contentLength) === 0) {
+			throw new FetchPostDownloadError('Response is empty');
+		}
```

Comment:
Ah, yes, that is the place I was telling you reminded me of the refactored `fetchJSON` in another PR and the validation of payload to make sure it's non-empty (or in [CLIM-1088](https://github.com/CanadianClimateDataPortal/climatedata-wp-theme/pull/640/files#r2589935043))